### ISSUE TITLE
FileAlreadyExistsException when using multipartFormData body parser

### ISSUE THREAD

**Issue by karimagnusson** [playframework/playframework#7700]
#### JDK 1.8.0_141

I am using multipartFormData body parser to upload a file. When I run multiple upload requests at the same time I regularly get these two errors.
`FileAlreadyExistsException`
`DirectoryNotEmptyException`

Example:
` Execution exception[[FileAlreadyExistsException: /var/folders/1t/cwyz6701029fmqwqyy4_200r0000gn/T/playtemp3860699828886373735]]`
It seems that creation of temporary files is not thread safe.

**Comment by gmethvin** [comment]
@karimagnusson What's the full stack trace of the error?

**Comment by karimagnusson** [comment]
### Stack trace for FileAlreadyExistsException:

! @7511lhil7 - Internal server error, for (POST) [/video/resumable/upload/j64y8tnk?resumableChunkNumber=52&resumableChunkSize=1048576&resumableCurrentChunkSize=1048576&resumableTotalSize=72968045&resumableType=video%2Fmp4&resumableIdentifier=72968045-f8fa54595d404450a487ce1f08012574m4v&resumableFilename=f8fa54595d404450a487ce1f08012574.m4v&resumableRelativePath=f8fa54595d404450a487ce1f08012574.m4v&resumableTotalChunks=69] ->
 
play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[FileAlreadyExistsException: /var/folders/1t/cwyz6701029fmqwqyy4_200r0000gn/T/playtemp4016054458618813336]]
	at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:255)
	at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:180)
	at play.core.server.AkkaHttpServer$$anonfun$13$$anonfun$apply$1.applyOrElse(AkkaHttpServer.scala:252)
	at play.core.server.AkkaHttpServer$$anonfun$13$$anonfun$apply$1.applyOrElse(AkkaHttpServer.scala:251)
	at scala.concurrent.Future$$anonfun$recoverWith$1.apply(Future.scala:344)
	at scala.concurrent.Future$$anonfun$recoverWith$1.apply(Future.scala:343)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
	at play.api.libs.streams.Execution$trampoline$.execute(Execution.scala:70)
	at scala.concurrent.impl.CallbackRunnable.executeWithValue(Promise.scala:40)
	at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:248)
Caused by: java.nio.file.FileAlreadyExistsException: /var/folders/1t/cwyz6701029fmqwqyy4_200r0000gn/T/playtemp4016054458618813336
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:88)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
	at sun.nio.fs.UnixFileSystemProvider.createDirectory(UnixFileSystemProvider.java:384)
	at java.nio.file.Files.createDirectory(Files.java:674)
	at java.nio.file.Files.createAndCheckIsDirectory(Files.java:781)
	at java.nio.file.Files.createDirectories(Files.java:727)
	at play.api.libs.Files$DefaultTemporaryFileCreator.create(Files.scala:163)
	at play.core.parsers.Multipart$$anonfun$handleFilePartAsTemporaryFile$1.apply(Multipart.scala:121)
	at play.core.parsers.Multipart$$anonfun$handleFilePartAsTemporaryFile$1.apply(Multipart.scala:119)

### Stack trace for DirectoryNotEmptyException:

! @7511m7i2g - Internal server error, for (POST) [/video/resumable/upload/j64yewa1?resumableChunkNumber=9&resumableChunkSize=1048576&resumableCurrentChunkSize=1048576&resumableTotalSize=32203414&resumableType=video%2Fquicktime&resumableIdentifier=32203414-f854d2ed4e314292875ad2f2b7bcca96mov&resumableFilename=f854d2ed4e314292875ad2f2b7bcca96.mov&resumableRelativePath=f854d2ed4e314292875ad2f2b7bcca96.mov&resumableTotalChunks=30] ->
 
play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[DirectoryNotEmptyException: /var/folders/1t/cwyz6701029fmqwqyy4_200r0000gn/T/playtemp5242047326278202622]]
	at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:255)
	at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:180)
	at play.core.server.AkkaHttpServer$$anonfun$13$$anonfun$apply$1.applyOrElse(AkkaHttpServer.scala:252)
	at play.core.server.AkkaHttpServer$$anonfun$13$$anonfun$apply$1.applyOrElse(AkkaHttpServer.scala:251)
	at scala.concurrent.Future$$anonfun$recoverWith$1.apply(Future.scala:344)
	at scala.concurrent.Future$$anonfun$recoverWith$1.apply(Future.scala:343)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
	at play.api.libs.streams.Execution$trampoline$.execute(Execution.scala:70)
	at scala.concurrent.impl.CallbackRunnable.executeWithValue(Promise.scala:40)
	at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:248)
Caused by: java.nio.file.DirectoryNotEmptyException: /var/folders/1t/cwyz6701029fmqwqyy4_200r0000gn/T/playtemp5242047326278202622
	at sun.nio.fs.UnixFileSystemProvider.implDelete(UnixFileSystemProvider.java:242)
	at sun.nio.fs.AbstractFileSystemProvider.deleteIfExists(AbstractFileSystemProvider.java:108)
	at java.nio.file.Files.deleteIfExists(Files.java:1165)
	at play.api.libs.Files$DefaultTemporaryFileCreator.play$api$libs$Files$DefaultTemporaryFileCreator$$playTempFolder(Files.scala:181)
	at play.api.libs.Files$DefaultTemporaryFileCreator.create(Files.scala:163)
	at play.core.parsers.Multipart$$anonfun$handleFilePartAsTemporaryFile$1.apply(Multipart.scala:121)
	at play.core.parsers.Multipart$$anonfun$handleFilePartAsTemporaryFile$1.apply(Multipart.scala:119)
	at play.core.parsers.Multipart$$anonfun$multipartParser$1$$anonfun$6.apply(Multipart.scala:79)
	at play.core.parsers.Multipart$$anonfun$multipartParser$1$$anonfun$6.apply(Multipart.scala:77)
	at akka.stream.impl.fusing.MapAsync$$anon$23.onPush(Ops.scala:1172)

**Comment by marcospereira** [comment]
Thank you, @karimagnusson. I can see that both methods where the exceptions are originating have a clear documentation that they can fail. [`Files.createDirectories` documentation says](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#createDirectories-java.nio.file.Path-java.nio.file.attribute.FileAttribute...-):

> FileAlreadyExistsException - if dir exists but is **not a directory** (optional specific exception)

While this is pretty unusual, that is something that we need to handle. But it would be good if you can create an example project to reproduce this problem because, as you can see in the docs, the exception is thrown when the path **is not a directory**. I couldn't find a situation where we are creating a file (or symlink) instead of a directory.

Is this just using the regular file upload?

The second exception is something we need to fix without a doubt as it is documented in [`Files.deleteIfExists`](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#deleteIfExists-java.nio.file.Path-):

> **Throws:**
DirectoryNotEmptyException - if the file is a directory and could not otherwise be deleted because the directory is not empty (optional specific exception)

I'm working in a fix for that.

**Comment by marcospereira** [comment]
Just to be clearer, my main concern about `FileAlreadyExistsException` is that if we are somehow creating a file instead of a directory, we will eventually move the problem to somewhere else. That is why I consider that would be good to have a reproducer here.

**Comment by richdougherty** [comment]
I checked the implementation of `Files.createDirectories` locally and it does ignore duplicate directories, so it shouldn't throw a `FileAlreadyExistsException`:

```scala
scala> val p = java.nio.file.Paths.get("/tmp/foo")
p: java.nio.file.Path = /tmp/foo
scala> java.nio.file.Files.createDirectories(p)
res2: java.nio.file.Path = /tmp/foo
scala> java.nio.file.Files.createDirectories(p)
res3: java.nio.file.Path = /tmp/foo
```

Here's the code in `java.nio.files.Files` which is used by `createDirectories` and which ignores directories:

```scala
private static void createAndCheckIsDirectory(Path dir,
                                              FileAttribute<?>... attrs)
    throws IOException
{
    try {
        createDirectory(dir, attrs);
    } catch (FileAlreadyExistsException x) {
        if (!isDirectory(dir, LinkOption.NOFOLLOW_LINKS))
            throw x;
    }
}
```

So I'm not sure how we get this exception here!

**Comment by richdougherty** [comment]
Hi @karimagnusson, a few more questions for you: :)
- how often does this happen? how many concurrent requests are you running?
- do you have a simple project or code that we can use to reproduce this exception?
- any chance there's a symlink in this path? (https://bugs.openjdk.java.net/browse/JDK-8130464)

**Comment by karimagnusson** [comment]
Hi, thanks for a quick response.

In my project I am using [resumablejs](http://www.resumablejs.com/) to upload large files in chunks. By default there are 3 chunks uploaded concurrently. When I upload a file of 70MB I get one or two errors most of the time.

I could not find any symlinks in the JVM temporary folder where the play temp folders are.

I created a simple demo to reproduce the error.
[upload-demo.zip](https://github.com/playframework/playframework/files/1214991/upload-demo.zip)
In application.conf you will have to set a path for the uploaded files.
`video.inbox = "/path/to/folder/"`
Then start the server and go to http://localhost:9000/assets/upload/index.html
There is no UI, just check the console.

If you do not get errors you can try to upload a larger file or change `simultaneousUploads` in index.html, line 47.

**Comment by richdougherty** [comment]
@karimagnusson, thanks a lot. We'll check that out.

**Comment by marcospereira** [comment]
@karimagnusson, see the current status of #7713.

Thank you so much for helping us to better identify the problem.

**Comment by richdougherty** [comment]
Hi @karimagnusson, we've just released Play 2.6.3 which should resolve this issue. If you get a chance, it would be great if you could test it out. :)

Thanks a lot for your help with reproducing the problem. 🍨

**Comment by karimagnusson** [comment]
Hi, I tried Play 2.6.3 and it works like a charm, no errors.
I am happy to have been able to contribute to the solution of the problem, thanks for a quick response.
